File: examples/lwjgl-map/src/main/java/org/maplibre/nativeffi/examples/lwjglmap/MetalBorrowedTexture.java
package org.maplibre.nativeffi.examples.lwjglmap;

import static org.lwjgl.system.MemoryUtil.NULL;

import org.maplibre.nativeffi.render.NativePointer;

final class MetalBorrowedTexture implements AutoCloseable {
  private final MetalContext context;
  private long texture;

  MetalBorrowedTexture(MetalContext context, Viewport viewport) {
    this.context = context;
    this.texture = context.createBorrowedTexture(viewport);
  }

  NativePointer pointer() {
    return NativePointer.ofAddress(texture);
  }

  long texture() {
    return texture;
  }

  @Override
  public void close() {
    if (texture != NULL) {
      context.releaseObject(texture);
      texture = NULL;
    }
  }
}

File: examples/lwjgl-map/src/main/kotlin/org/maplibre/nativeffi/examples/lwjglmap/MetalBorrowedTexture.kt
package org.maplibre.nativeffi.examples.lwjglmap

import org.lwjgl.system.MemoryUtil.NULL

internal class MetalBorrowedTexture(graphicsContext: GraphicsContext, viewport: Viewport) :
  AutoCloseable {
  private val metalContext = graphicsContext as MetalContext
  private var texture: Long = metalContext.createBorrowedTexture(viewport)

  fun texture(): Long = texture

  override fun close() {
    if (texture != NULL) {
      metalContext.releaseObject(texture)
      texture = NULL
    }
  }
}
